tools/diagnostics/full_bundle.py

```diff
@@ -0,0 +1,181 @@
+
+
+def to_dcos_service_name(service_name: str) -> str:
+    """DC/OS service names don't contain the first slash.
+    i.e.: |     SDK service name     |   DC/OS service name    |
+          |--------------------------+-------------------------|
+          | /data-services/cassandra | data-services/cassandra |
+    """
+    return service_name.lstrip("/")
+
+
+def is_service_named(service_name: str, service: dict) -> bool:
+    return service.get("name") == to_dcos_service_name(service_name)
+
+
+def is_service_active(service: dict) -> bool:
+    return service.get("active") is True
+
+
+def services_with_name(service_name: str, services: List[dict]) -> List[dict]:
+    return [s for s in services if is_service_named(service_name, s)]
+
+
+def active_services_with_name(service_name: str, services: List[dict]) -> List[dict]:
+    return [s for s in services if is_service_named(service_name, s) and is_service_active(s)]
+
+
+def is_service_scheduler_task(package_name: str, service_name: str, task: dict) -> bool:
+    labels = task.get("labels", [])
+    dcos_package_name = next(
+        iter([l.get("value") for l in labels if l.get("key") == "DCOS_PACKAGE_NAME"]), ""
+    )
+    dcos_service_name = next(
+        iter([l.get("value") for l in labels if l.get("key") == "DCOS_SERVICE_NAME"]), ""
+    )
+    return dcos_package_name == package_name and dcos_service_name == to_dcos_service_name(
+        service_name
+    )
+
+
+def directory_date_string() -> str:
+    return date.strftime(datetime.utcnow(), "%Y%m%dT%H%M%SZ")
+
+
+class FullBundle(Bundle):
+    def __init__(self, package_name, service_name, bundles_directory):
+        self.package_name = package_name
+        self.service_name = service_name
+        self.bundles_directory = bundles_directory
+        self.output_directory = self._create_bundle_directory()
+
+    def _configure_logging(self):
+        """Configures logging to write script output to bundle output directory.
+        """
+        logging.basicConfig(
+            format="%(asctime)-15s %(levelname)s %(message)s",
+            level="INFO",
+            handlers=[
+                logging.FileHandler(os.path.join(self.output_directory, "script.log")),
+                logging.StreamHandler(),
+            ],
+        )
+
+    def _bundle_directory_name(self) -> str:
+        _, cluster_name, _ = sdk_cmd.run_cli("config show cluster.name", print_output=False)
+        return "{}_{}_{}".format(
+            cluster_name,
+            sdk_utils.get_deslashed_service_name(self.service_name),
+            directory_date_string(),
+        )
+
+    def _create_bundle_directory(self) -> str:
+        directory_name = os.path.join(self.bundles_directory, self._bundle_directory_name())
+
+        if not os.path.exists(directory_name):
+            log.info("Creating directory %s", directory_name)
+            os.makedirs(directory_name)
+
+        return directory_name
+
+    def create(self) -> (int, "FullBundle"):
```

Comment:
Is this how one specifies a class return?

Reply:
[Yup](https://www.python.org/dev/peps/pep-0484/#forward-references).

Reply:
Couldn't we use it without the quotation marks here?

Reply:
No, because `FullBundle` still isn't defined at this point. The link above has more detail.